hs-abci-sdk/src/Tendermint/SDK/BaseApp/Metrics/Prometheus.hs

```diff
@@ -0,0 +1,187 @@
+      a <- runT action
+      end <- liftIO $ getCurrentTime
+      let time = fromRational . (* 1000.0) . toRational $ (end `diffUTCTime` start)
+      evalMetrics state (observeHistogram histName time)
+      actionRes <- raise $ evalMetrics state a
+      pure $ (, time) <$> actionRes
+    )
+
+evalWithMetrics
+  :: Member (Embed IO) r
+  => Member (Reader MetricsState) r
+  => Sem (Metrics ': r) a
+  -> Sem r a
+evalWithMetrics action = ask >>= flip evalMetrics action
+
+--------------------------------------------------------------------------------
+-- indexes @NOTE: maybe clean this up
+--------------------------------------------------------------------------------
+
+countToIdentifier :: CountName -> MetricIdentifier
+countToIdentifier (CountName name labels) = MetricIdentifier
+  { metricIdName = fixMetricName name
+  , metricIdLabels = MetricId.fromList labels
+  , metricIdHistoBuckets = []
+  }
+
+histogramToIdentifier :: HistogramName -> MetricIdentifier
+histogramToIdentifier (HistogramName name labels buckets) = MetricIdentifier
+  { metricIdName = fixMetricName name
+  , metricIdLabels = MetricId.fromList labels
+  , metricIdHistoBuckets = buckets
+  }
+
+-- | Prometheus registry index key
+mkPrometheusMetricId :: MetricIdentifier -> MetricId.MetricId
+mkPrometheusMetricId MetricIdentifier{..} =
+  MetricId.MetricId (MetricId.Name metricIdName) metricIdLabels
+
+-- | Index key for storing metrics
+metricIdStorable :: MetricIdentifier -> (Text, MetricId.Labels)
+metricIdStorable c = (fixMetricName $ metricIdName c, fixMetricLabels $ metricIdLabels c)
+  where fixMetricLabels =
+          MetricId.fromList .
+          map (fixMetricName *** fixMetricName) .
+          MetricId.toList
+
+fixMetricName :: Text -> Text
+fixMetricName = Text.map fixer
+  where fixer c = if c `elem` validChars then c else '_'
+        validChars = ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9'] ++ "_"
+
+-- | Core metrics state
+type MetricsMap a = Map (Text, MetricId.Labels) a
+data MetricsState = MetricsState
+  { metricsRegistry   :: Registry.Registry
+  , metricsCounters   :: MVar (MetricsMap Counter.Counter)
+  , metricsHistograms :: MVar (MetricsMap Histogram.Histogram)
+  }
+
+emptyState :: IO MetricsState
+emptyState = do
+  counters <- newMVar Map.empty
+  histos <- newMVar Map.empty
+  registry <- Registry.new
+  return $ MetricsState registry counters histos
+
+-- | Intermediary prometheus registry index key
+data MetricIdentifier = MetricIdentifier
+  { metricIdName         :: Text
+  , metricIdLabels       :: MetricId.Labels
+  , metricIdHistoBuckets :: [Double]
+  }
+
+instance IsString MetricIdentifier where
+  fromString s = MetricIdentifier (fromString s) mempty mempty
+
+--------------------------------------------------------------------------------
+-- setup @TODO: add error handling
+--------------------------------------------------------------------------------
+
+getEnvVarBoolWithDefault :: MonadIO m => String -> Bool -> m Bool
```

Comment:
This shouldn't really be here. You should only use environment variables in application code.  for example this

```
  shouldStart <- getEnvVarBoolWithDefault "STATS_ENABLED" True
  if not shouldStart
    then return Nothing
    else do
      bindPort <- readEnvVarWithDefault "STATS_PORT" 9200

```

means that this function should take a `Maybe Port` and figure out what to do based on `Nothing | Just Port` and then shove the `Maybe Port` as part of the metrics config. Then in application code you can use an EnvVar to discover the value.

Reply:
Does that have to be part of the BaseApp context similar to logging? Or can it just live inside of the AppConfig?